src/python/pants/backend/python/subsystems/pex_build_util.py

```diff
@@ -312,19 +372,98 @@ def _prepare_inits(self) -> Set[str]:
         ns_package.write(b'__import__("pkg_resources").declare_namespace(__name__)  # type: ignore[attr-defined]')
         ns_package.flush()
         for missing_init_file in missing_init_files:
+          self._all_added_sources_resources.append(Path(missing_init_file))
           self._builder.add_source(filename=ns_package.name, env_filename=missing_init_file)
     return missing_init_files
 
   def set_emit_warnings(self, emit_warnings):
     self._builder.info.emit_warnings = emit_warnings
 
+  def _shuffle_underlying_pex_builder(self, new_builder: PEXBuilder) -> Tuple[PexInfo, Path]:
+    orig_info = self._builder.info.copy()
+    orig_chroot = self._builder.chroot()
+
+    self._builder = new_builder
+
+    return (orig_info, Path(orig_chroot.path()))
+
+  def _shuffle_original_build_info_into_ptex(self):
+    """Create a "dehydrated" ptex file without any of its requirements, and specify that in two *-INFO files.
+
+    See pex_bootstrap/_ptex_launcher.py for details of how these files are used.
+    """
+    # Replace the original builder with a new one, and just pull files from the old chroot.
+    ipex_info, orig_chroot = self._shuffle_underlying_pex_builder(
+      PEXBuilder(interpreter=self._builder.interpreter))
+
+    # PTEX-INFO: Contains resolver settings and source files to copy over into the dehydrated .ipex
+    #            file.
+    all_code = [str(src) for src in self._all_added_sources_resources]
+    prefixed_code_paths = [os.path.join(APP_CODE_PREFIX, src) for src in all_code]
+    for src, prefixed in zip(all_code, prefixed_code_paths):
+      # NB: Need to add under 'source' label for `self._prepare_inits()` to pick it up!
+      self._builder.chroot().copy(os.path.join(str(orig_chroot), src),
+                                  prefixed,
+                                  label='source')
+
+    python_repos = self._python_repos_subsystem
+
+    resolver_settings = dict(
+      indexes=list(python_repos.indexes),
+      find_links=list(self._all_find_links),
+    )
+
+    ptex_info = dict(
+      code=prefixed_code_paths,
+      resolver_settings=resolver_settings,
+      # This setting is consumed in ptex.py to determine whether to pyc-compile sources when
+      # building the .ipex. Because .ipex files also only contain requirements for a single
+      # platform and interpreter, there shouldn't be any expectation that they will run on a
+      # different interpreter, so pyc-compiling here is probably safe to do.
+      bytecode_compile=True,
```

Comment:
We currently hard-code this everywhere else we set it in Pants to `False` to get deterministic output. I'd suggest doing the same here, unless there's a good reason not to?

Reply:
That sounds great!